### DISCUSSION TITLE
Wired data loss issue

### DISCUSSION THREAD

**Discussion by mahirchavda** [pocketbase/pocketbase#2383] [Q&A] [answered]
Today some of collection are not showing in the UI that were created in last 2-3 weeks. I'm not sure what happened.

However, I see the migration files present in the pb_migrations folder and I believe those are the changes that I lost for unknown reason. 

Is it possible to recover those changes (with collection records)? and find the root cause?

Attaching screenshots

PS. 
This happened second time with me and It happens couple of weeks after upgrading pocketbase (my observation)

I'm running pocketbase on raspberry pi using docker. pb_data folder is mounted as ./pb_data:/pocketbase/pb_data
![image](https://user-images.githubusercontent.com/25617583/235290067-facc9cba-dc04-46c0-a091-169f942f6fc2.png)
![image](https://user-images.githubusercontent.com/25617583/235290074-35d4108c-5924-42bf-b66b-5c696f3ec78a.png)
![image](https://user-images.githubusercontent.com/25617583/235290078-1f1b0730-67d4-43ed-957c-1a692a48cbf2.png)

**Comment by ganigeorgiev** [comment]
This is most likely similar to https://github.com/pocketbase/pocketbase/discussions/1558.

Make sure that the `pb_data` volume is properly mounted.

> However, I see the migration files present in the pb_migrations folder and I believe those are the changes that I lost for unknown reason.

`pb_migrations` contains the changes of your collections configuration (eg. create/update of a collection or field). All unapplied migrations will be automatically applied on `serve`.

If you are still not able to identify the isse, please provide a little more details about your setup (eg. Dockerfile, docker-compose, etc.) so that I can try to reproduce it.

**Comment by mahirchavda** [comment]
Thanks for checking. I'm not sure it will be reproduced or not. It seems more like setup issue but any help would be really appreciated.

To update the pocketbase, I update PB_version from docker-compose file and then run `docker compose up -d --build` comamnd

Dockerfile
```
FROM alpine:3.16.3

WORKDIR /pocketbase

ARG PB_VERSION

RUN apk add --no-cache \
    unzip

# download and unzip PocketBase
ADD https://github.com/pocketbase/pocketbase/releases/download/v${PB_VERSION}/pocketbase_${PB_VERSION}_linux_arm64.zip /pocketbase/pb.zip

RUN unzip /pocketbase/pb.zip -d /pocketbase/

# start PocketBase
CMD ["/pocketbase/pocketbase", "serve", "--http=0.0.0.0:8090"]
```


docker-compose.yml
```
version: "3"
services: 
  pocketbase:
    container_name: pocketbase
    build: 
      context: .
      args:
        - PB_VERSION=0.15.0
    volumes:
      - ./pb_data:/pocketbase/pb_data
    ports:
      - "8090:8090"
    restart: unless-stopped
```

**Reply by ganigeorgiev** [reply, marked as answer]
I'm not able to reproduce the issue with the provided docker configuration.

Make sure that your local `pb_data` is persistent and that there is no other application that could overwrite it.

_Also note that in the above setup you are not mounting anywhere the `pb_migrations` directory._

**Comment by mahirchavda** [comment]
One thing I would like to understand is why migration files are getting created automatically and not applied directly to database ? If that helps in any way in debugging the issue.

**Reply by ganigeorgiev** [reply]
> One thing I would like to understand is why migration files are getting created automatically and not applied directly to database ?

They are applied to the database and the applied migration filenames are stored in the `_migrations` table. The `pb_migrations` files contains "instructions" what changes to apply and are there to commit them in your repo, so that your other team members or when you want to start your project from scratch, to have all collections created with the application start.

You can disable the automigrations creation with the `--automigrate=0` flag.

**Comment by mahirchavda** [comment]
Is it possible to run migration realated js files to recover at least collection schema changes? I tried mounting pb_migraitons folder and ran `./pocketbase migrate` and `./pocketbase migrate up` but it's not working.

**Reply by ganigeorgiev** [reply]
I'm confused and I'm not sure that I understand your setup.
How did you generated the `pb_migrations` directory content if previously it wasn't mounted?
Was it created by running a local PocketBase executable outside of Docker?

In any case, if an explicit `--migrationsDir` path is not provided, the prebuilt executable will look for the `pb_migrations` one directory up relative to the `pb_data` location, aka. in your case the mount point should be something like: `./pb_migrations:/pocketbase/pb_migrations`.

Only the **unapplied** migrations will be executed automatically. Already applied migrations are skipped (aka. those that are listed in the `_migrations` table).

**Reply by mahirchavda** [reply]
You are right, I copied pb_migrations directory from the docker container and then created new container with pb_migrations mounted. those migrations are already present in the _migration table. All the recent 2-3 weeks change are lost today (except _migrations). I'll apply them manually now. Thanks for all the help

**Comment by philffm** [comment]
I also had a weird issue today that might be related to this one. One collection named "content" which I created last week for (rich) content of my web app just disappeared today (after restarting my docker instance). I only upgraded to the latest version last week. 

(new) fields and data of other collections however are still there which is why I assume no permission issue 🤷🏽‍♂️. I then created a new collection with a different name and that one survived the container restart so far but I'm wondering if it could happen again with more important collections

**Reply by ganigeorgiev** [reply]
I'm not sure how to help. I suspect that this is something related with the Docker setup (eg. misconfigured volume path or something similar). The collections changes could be also caused by a migration script (aka. if you use the automigrate functionality, do you mount the `pb_migrations` dir as part of your docker image?). In any case, if you manage to reproduce it, please provide the steps and I'll try to investigate it.

**Comment by ganigeorgiev** [comment]
Just in case someone stumbled on this discussion - the cause for the above reported issue is most likely the same as in https://github.com/pocketbase/pocketbase/discussions/2551#discussioncomment-5983139.
